_drafts/Firmware-encryption-with-python.md

```diff
@@ -0,0 +1,643 @@
+        print( CipherbytesLittle )
+        
+        bytes_so_far_cryted += BYTES_READ
+        bytes_remaining = BinaryFileSize - bytes_so_far_cryted
+        if( bytes_remaining <= 0 ): bytes_remaining = 0
+        print( bytes_remaining )
+        
+    PlainFirmware.CloseFile()
+    CipherFirmware.CloseFile()
+main()
+```
+## Firmware Decryption Implementation
+
+Our update system managed to receive the encrypted firmware now it all depends on our MCU we have to decrypt the firmware and install it in a section of flash memory where it can be executed by our bootloader. In this post we are going to use the AES Hardware Accelerator to perform firmware decryption. All the code for this post is written for the [B-L4S5I-IOT01A](https://www.st.com/en/evaluation-tools/b-l4s5i-iot01a.html#) Dev kit with STM32L4S5 MCU by ST Micro.
+
+### Define encryption keys
+
+Remember that AES is a symmetric key algorithm so it is no longer necessary to generate new keys to decrypt, we have to use the same keys used to encrypt. Let’s define the initialization vector and the key:
+
+```c
+static const uint8_t pKeyAES[16] __attribute__ ((aligned (4))) = {
+                            0x58,0x3b,0xf4,0x90,0x1f,0x62,0x35,0x10,0xa1,0x40,
+                            0x8e,0x8b,0xe7,0xbd,0x6f,0x5a};
+static const uint8_t NonceAES[16] __attribute__ ((aligned (4))) = {
+                            0xf8,0x24,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
+                            0x00,0x00,0x00,0x00,0x00};
+```
+
+### Decryption with AES hardware accelerator
+
+The AES hardware accelerator encrypts and decrypts data and supports CTR mode, for 128-bit or 256-bit key sizes. It is necessary to configure the peripheral according to how we encrypt the firmware to be sure of obtaining the same result when decrypting.
+
+So we need to make a configuration with these points:
+
+1. Decryption with chaining mode Counter (CTR) mode.
+2. Key sizes of 128 bits.
+
+Let’s see how the hardware accelerator processes a message when it works in CTR mode presenting a structure similar to the one we defined when encrypting:
+
+![](/img/fwup-encryption/MessgaeCTR.jpg)
+
+The structure of this message is:
+
+1. A 16-byte initial counter block (ICB), composed of two distinct fields:
+* Initialization vector (IV): a 96-bit value that must be unique for each encryption
+cycle with a given key.
+* Counter: a 32-bit big-endian integer that is incremented each time a block
+processing is completed. The initial value of the counter must be set to 1.
+2. The plaintext P is encrypted as ciphertext C, with a known length. This length can be
+non-multiple of 16 bytes, in which case a plaintext padding is required.
+
+### Setup AES hardware accelerator
+
+Now that we know the necessary parameters we can load the configuration in the `CR` register in addition to initializing the key `(AES_KEYRx (KEY))` and initialization vector in the corresponding `(AES_IVRx)` registers:
+
+```c
+void aes_setup( void ) {
```

Comment:
Love this! Can you add a reference the STM32 reference manual so that people can follow along with the register initialization here?